### DISCUSSION TITLE
RFC: preventing id collisions and refactoring...thoughts?

### DISCUSSION THREAD

**Discussion by wt** [maxcountryman/tower-sessions#171] [Ideas]
So, I have been working on banging out some experiments on how to change the code in order to prevent id generation collisions (after discussion on #169). Due to the need for a major version change for any of the real improvements, I think it would be better to engage with this in a more iterative fashion. With regard to changing and maintaining, I have run into the following serious problems with the code:

* unit test coverage is almost non-existent
* the files are just too damn big

So, I think this gives us an opportunity to do some pretty serious refactoring. You can see the very beginnings of where I am heading at [here](https://github.com/wt/tower-sessions/commits/prevent_id_collisions/). This branch is currently a PoC and not ready for merging. It's just to give a little structure to this conversation.

The important piece of those changes is that I am introducing testing infrastructure and completely new types called NewSessionStore and NewSession. These will be used to make any incompatible changes with the current SessionStore and Session. I will try to factor as much code out of Session and SessionStore so that it is used by the old and the new versions. Once that is in progress, I am hoping a more concrete plan for making the major version change can be formed.

As it stands, I will probably hide the new stuff behind a feature flag that we can make not default until the major version change. This is not done in the branch, I will probably modify the branch to include a commit adding the feature before adding the types.

The next steps I plan on taking are literally breaking up some of the files (especially tower-session-core/src/session.rs).

I am also pretty uncomfortable with the logic that's currently embedded into the Session impl. I am hoping we can zhuzh that around a bit to make testing easier.

Does anyone, especially @maxcountryman, have any thoughts about this?

Thanks! wt

**Comment by maxcountryman** [comment]
Thanks for working on this! I appreciate you taking an interest in this tricky problem.

I've looked over the branch you mentioned and here are a few quick thoughts:

At a high level I would prefer we not introduce a new type and instead rip the bandage off. Because we're pre-1.0, we have more leeway here to make breaking changes. I'd also rather introduce breaking changes than introduce a second set of types which need to be maintained and will ultimately need to be removed anyway.

On a more granular level, I'm not sure the OnceCell approach will work and more specifically we will want to ensure that we are not removing the trace logs in the process: the suspicious activity warning is required by OWASP, for instance. Moreover the semantic of Option regarding the record is important and so either it shouldn't be changed or if it is it needs to be done so carefully so as not to change the current semantic.

> unit test coverage is almost non-existent

Just to make sure we're on the same page here, we have around 70% coverage, with doctests providing unit-style tests of the session interface (these are run by CI). I'm not opposed to adding more tests, extending tests by adding a test mod, etc but I did want to make sure you had seen those.

Regarding testing, I think it could make sense to introduce tests independently from the interface changes: those we can merge right away and continue to extend as the interface is evolved to address the collision problem.

Please let me know how I can be helpful. 😁

**Reply by wt** [reply]
With regard to testing, I didn't even count the doctests. Sorry about that.

**Reply by wt** [reply]
I'm happy to know we can evolve the interface a bit faster due to the pre-1.0 status.

In that vein, I am happy to make these changes to the types directly instead of building the separate types. I was just trying to respond to your feedback about the evolution of the interface.

On submitting testing separate, I added some of the testing as the early commits as I was considering submitting those separately.

I have some comments about your suggestion on the trait below that I think we should hash out.

**Reply by wt** [reply]
I went ahead and submitted a PR for the test I wrote for the expiry time generation. This has nothing to do with the changes of the types for the store (and any related changes).

**Reply by wt** [reply]
So, I gave some thought to the OnceCell method I started implementing. I do think it can work as long as we say that only a mutable session can have an id changed.

I have an implementation that should allow something like that following for a known new session creation:

```
let session = Session::new(session_store, expiry); //notice there is no id here

// Takes a &self and causes creating a new session in the store since the id has
// never been set. Also, record attr will be set to include no data in the session.
info!("session: {:}", session.id());

// doesn't cause a load from the store as the record was loaded before 
info!("record: {:?}", session.record());
```

Interior mutability would be needed to update the record data, like adding new keys for the session data, so you'd probably really want a `let mut session` in the example above to allow mutating the data in the session.

For a load of an existing session, it would look like this:
```
let session = Session::new(session_store, expiry);

// tries to load the record and id from the data store. If successful, sets the id
// and record attrs. If not, returns an error to be handled. This could be an error
// indicating a retryable error (e.g., couldn't talk to DB) or a unretryable error
// (e.g., session doesn't exist in store or maybe session is expired in store). Proper
// OWASP logging can happen here as it does today.
try_from_existing_id(&id)?;

// assuming no error from above; this will cause no hit to the store since that
// was already loaded during the try_from_existing_id call above
info!("session: {:}", session.id());

 // again, no hit to the store
info!("record: {:?}", session.record());
```

It's possible this cannot work, but I am part of the way there at this point. This experiment has not been reflected in the branch linked above. It's not ready for sharing yet.

**Reply by maxcountryman** [reply]
I did update the interior mutability impl to use a MappedMutexGuard such that we don't need to return Option from get_record anymore, which does improve ergonomics and clarity imo.

If OnceCell could somehow make the interface over Record sync that would be nice but I would think that's not possible so I'm not exactly sure what the benefit of it would be?

**Reply by wt** [reply]
I'll look into that. I hadn't really settled the type for the record attr yet. I just needed something there. I need to finish the experiment.

**Reply by wt** [reply]
And FWIW, once cell is probably not the right type for the record attr. It's quite nice for the id attr.

**Comment by maxcountryman** [comment]
A thought I had here to help ease the transition to a new method could look something like this:

```rust
#[async_trait]
pub trait SessionStore: Debug + Send + Sync + 'static {
   /// A method for saving a session in a store.
    async fn create(&self, session_id: &Id) -> Result<()> {
        self.default_create(session_id).await
    }

    #[deprecated(note = "Please implement the `create` method directly. This default implementation will be removed in the future.")]
    async fn default_create(&self, session_id: &Id) -> Result<()> {
        self.save(session_id).await
    }

    /// A method for saving a session in a store.
    async fn save(&self, session_record: &Record) -> Result<()>;

    /// A method for loading a session from a store.
    async fn load(&self, session_id: &Id) -> Result<Option<Record>>;

    /// A method for deleting a session from a store.
    async fn delete(&self, session_id: &Id) -> Result<()>;
}
```

**Reply by wt** [reply]
I thought about doing something like that at first. However, it didn't sit right with me to leave the return type for load as `Result<Option<Record>>`. I think it needs to be changed to `Result<Record>` since not being able to load a record is really an error that needs to be handled.

My thought is this, the create method should atomically create a new session in the store. Any time a load happens and fails, it seems like that is an error that should be handled by creating a new session.

Also, I think that the create method should return the record for efficiency. When I create a new session, I should be able to return the record so that I don't have to do a separate load.

**Reply by maxcountryman** [reply]
Here's a somewhat improved version of this idea:

```rust

#[async_trait]
pub trait SessionStore: Debug + Send + Sync + 'static {
    /// A method for creating a session in a store.
    async fn create(&self, session_record: &Record) -> Result<Record> {
        default_create(self, session_record).await
    }

    /// A method for saving a session in a store.
    async fn save(&self, session_record: &Record) -> Result<()>;

    /// A method for loading a session from a store.
    async fn load(&self, session_id: &Id) -> Result<Option<Record>>;

    /// A method for deleting a session from a store.
    async fn delete(&self, session_id: &Id) -> Result<()>;
}

#[deprecated(
    note = "In order to prevent ID collisions, `SessionStore::create` must be implemented \
            directly. This default method will be removed in future releases."
)]
async fn default_create<S: SessionStore + ?Sized>(
    store: &S,
    session_record: &Record,
) -> Result<Record> {
    store.save(session_record).await?;
    Ok(session_record.clone())
}
```

There is a problem with using the deprecation warning here however: the message appears regardless of whenever the create method is implemented.

**Reply by maxcountryman** [reply]
> However, it didn't sit right with me to leave the return type for load as Result<Option<Record>>. I think it needs to be changed to Result<Record> since not being able to load a record is really an error that needs to be handled.

The idea here is that stores ultimately decide if a session is valid or not: they own this semantic canonically. 

So for example, a store might invalidate sessions for any arbitrary reason (our application may want to force users to reauthenticate, etc). However, the session manager is only eventually consistent with this semantic and so it's valid to ask for a session via ID and have the store return `None` since it may be the case that the user agent has a session which hasn't expired (in accordance with `Max-Age`) but that the store and thereby application no longer accepts as valid. 

This isn't an erroneous state so I don't think it would make sense to return an error and regardless we need to handle this state in a very specific way: by generating a new session.

**Reply by wt** [reply]
Not finding a session in the store does seem like an erroneous state to me. In my mind, the most sensible thing to do is to return an error so that whatever (e.g. the tower layer) is building the session object can handle it, perhaps by trying to create a session instead of loading an old one.

However, it never seems reasonable to proceed with a return of none for the session record. It seems like none always has to be handled. Even in your description that seems to be the case. Is there ever a case where it's okay to proceed with none.

Also, in my design above, I changed the session object building to make trying to load a specific session id an explicit step that can fail and be handled on its own. It's not an implicit part of the construction. This allows more explicit handling of that problem than the current design of the session object. And I think it allows the session store to be more precise and not return a none for the load.

The problem with the current design of session is that the building (the new method) must succeed, which is why it implicitly creates a new session id when needed. I'm my changed world. You construct the object, then explicitly attempt to load. That could return and error to be handled. The means that the builder of the session object can now implement the error handling however. You no longer have to roll it into the session object construction method.

I don't like the current method as any error handling, including number of retries for an operation, is buried in the session object logic. The session object should allow what's using it to make those decisions.

**Reply by maxcountryman** [reply]
So with `create` add to the trait, we'd essentially make a change to `Session::save` that would look something like this (this would be updated to use `MappedMutexGuard`, so that version wouldn't need to pattern match on `get_record`):

```rust
    pub async fn save(&self) -> Result<()> {
        if let Some(record) = self.get_record().await?.as_ref() {
            let mut must_create = false;

            {
                let mut session_id_guard = self.session_id.lock();
                if session_id_guard.is_none() {
                    *session_id_guard = Some(record.id);
                    must_create = true;
                }
            }

            if must_create {
                self.store.create(record).await?;
            } else {
                self.store.save(record).await?;
            }
        }
        Ok(())
    }
```

Using memory store as an example, an implementation of `create` could look like:

```rust
    async fn create(&self, record: &Record) -> session_store::Result<Record> {
        let mut record = record.clone();
        loop {
            let mut guard = self.0.lock().await;

            if let Entry::Vacant(e) = guard.entry(record.id) {
                e.insert((record.clone(), record.expiry_date));
                break;
            };

            // Assign a new ID since we otherwise would collide.
            record.id = Id::default();
        }
        Ok(record)
    }
```

Here stores have full control over deciding how to handle collisions and the session interface is updated to ensure that the semantic of the session ID dictates when `create` is used.

**Reply by maxcountryman** [reply]
It's probably easier to look at something a little more concrete, so [I fleshed this out a bit](https://github.com/maxcountryman/tower-sessions/compare/id-collision-mitigation?expand=1).

I have to say, I'm not too unhappy with this: it's a smaller change than I think it first seemed was required and I like the fact the `create` method is clear about taking exclusive ownership of the record (which also makes its usage cleaner).

**Reply by wt** [reply]
I'm not crazy about the store handling the conflict resolution. It feels similar to the problem of bufferbloat. I need to think about this.

My initial concerns are that it puts unknown load on the store due to the possibility of the transparent retries. There is probably a small chance of that due to the 2^128 space of the id.

However, hiding that kind of problem from the service creating the new session and delivering it to the user seems bad. I really can't shake the feeling that he user of the session object should be making decision about retry policies. For example, there may be cases where I would never want to retry and would prefer to just return a 5xx error to the client. I would think that wouldn't be most use cases, but that possibility is gone with your design. Also, you design retries as fast as possible. Should be there some kind of exponential backoff?

I'm really worried about something like this at scale that tries to hide such errors. I've been bitten by many systems that try to transparently retry without notifying the upper layers that such a thing is happening. There is almost always some unintentional destructive behavior that eventually emerges. My general inclination is to fail fast and let the upper layers choose how and when to retry.

And to be more concrete, I think the SessionManager layer should be the place where the retry logic lives, and the SessionLayerManager should be where the policy for the logic is configured. And have something like the following when initializing it:

```rust
let session_store = BlahSessionStore(...);
SessionManagerLayer::new(session_store)
    .with_same_site(SameSite::Lax)
    .with_retry_policy(RetryPolicy::builder()
        .with_id_collision_retries(3)
        .build())
```

Honestly, I think that being able to set the policies like that will also make the objects much easier to test since the policy is separated from the logic in a cleaner way.

Having said that, I think that your design will be better than where it stands today.

**Reply by wt** [reply]
This article seems somehow relevant: https://www.maxcountryman.com/articles/let-it-fail

**Reply by maxcountryman** [reply]
Thank you for sharing your thoughts and concerns. I appreciate the depth of your analysis and your commitment to improving the project.

The design direction you suggest is indeed one possibility. However, it diverges from the core philosophy of this crate, which leans towards leaving such decisions to the discretion of store implementers. That said, I do see the value in your approach, and I can envision a version of `BlahSessionStore` that provides the interface you've described:

```rust
let session_store = BlahSessionStore(...)
    .with_retry_policy(
        RetryPolicy::builder()
            .with_id_collision_retries(3)
            .build()
    );
let session_layer = SessionManagerLayer::new(session_store)
    .with_same_site(SameSite::Lax);
```

This approach allows session stores, which are types we control, to offer extended interfaces that cater to specific needs. While defining an interface for store retry policies might be beyond the scope of this crate, it is certainly within the scope of `SessionStore` implementations. In fact, such opinionated implementations could be published as their own crates, allowing for greater flexibility and customization.

With all this in mind, my plan is to introduce the `create` method I've outlined on the branch above and move toward cutting a new release to ensure we're setting folks up for mitigation on a reasonable timeline.

Thank you again for reporting this issue and thinking through the design space. It's a tricky balance, especially considering this thing is already out in the wild with active users depending on our API.

**Reply by wt** [reply]
I like that idea. Allows the stores to implement whatever config they want. That seems like a relatively good idea. If you find there are common configurations, you could always add that to the interface later.

I think this is a good incremental step. Thanks for working with me on this.

**Comment by maxcountryman** [comment]
Regarding testing, I'm adding some additional service tests as we work on introducing signing and encryption over here: https://github.com/maxcountryman/tower-sessions/pull/157/commits/90eb541c0299f0ff157c040b86f686b309c5eb40

**Reply by wt** [reply]
Thanks for the pointer. I will check that out.
